compiler/rustc_typeck/src/check/intrinsic.rs

```diff
@@ -21,36 +21,68 @@ fn equate_intrinsic_type<'tcx>(
     tcx: TyCtxt<'tcx>,
     it: &hir::ForeignItem<'_>,
     n_tps: usize,
+    n_lts: usize,
     sig: ty::PolyFnSig<'tcx>,
 ) {
-    match it.kind {
-        hir::ForeignItemKind::Fn(..) => {}
+    let (gen_lts, gen_tys, gen_cns, span) = match &it.kind {
+        hir::ForeignItemKind::Fn(.., generics) => {
+            let mut gen_lts = 0;
+            let mut gen_tys = 0;
+            let mut gen_cns = 0;
+
+            for param in generics.params {
+                match param.kind {
+                    hir::GenericParamKind::Lifetime { .. } => {
+                        gen_lts += 1;
+                    }
+                    hir::GenericParamKind::Type { .. } => {
+                        gen_tys += 1;
+                    }
+                    hir::GenericParamKind::Const { .. } => {
+                        gen_cns += 1;
+                    }
+                }
+            }
+
+            (gen_lts, gen_tys, gen_cns, generics.span)
+        }
         _ => {
             struct_span_err!(tcx.sess, it.span, E0622, "intrinsic must be a function")
                 .span_label(it.span, "expected a function")
                 .emit();
             return;
         }
-    }
-
-    let i_n_tps = tcx.generics_of(it.def_id).own_counts().types;
-    if i_n_tps != n_tps {
-        let span = match it.kind {
-            hir::ForeignItemKind::Fn(_, _, ref generics) => generics.span,
-            _ => bug!(),
-        };
+    };
 
-        tcx.sess.emit_err(WrongNumberOfTypeArgumentsToInstrinsic {
+    if gen_lts != n_lts {
```

Comment:
It would be nice to abstract the following three conditions into a loop or something, because they all essentially have the same effect.

Reply:
I moved the check into a closure and used nested `if` statements to make sure only one error gets emitted at a time.